common/utilities/time/waiting-on.h

```diff
@@ -0,0 +1,134 @@
+
+    public:
+        wait_state_t() = default;   // allow default ctor
+        wait_state_t( T const & t )
+            : _value( t )
+        {
+        }
+
+        operator T &() { return _value; }
+        operator T const &() const { return _value; }
+
+        T* operator->() { return &_value; }
+        T const* operator->() const { return &_value; }
+
+        // Set a new value and signal
+        void signal( T const & t )
+        {
+            _value = t;
+            signal();
+        }
+        // Signal with the current value
+        void signal()
+        {
+            _cv.notify_one();
+        }
+    };
+private:
+    std::shared_ptr< wait_state_t > _ptr;
+
+    // When we declare the signalling lambda for the other thread, we need to pass it the weak_ptr.
+    // This class wraps it up nicely, so you can write:
+    //     waiting_on< bool > invoked( false )
+    //     auto thread_function = [invoked = invoked.in_thread()]() {
+    //         invoked.signal( true );
+    //     }
+    //
+public:
+    class in_thread_
+    {
+        std::weak_ptr< class wait_state_t > const _ptr;
+
+    public:
+        in_thread_( waiting_on const& local )
+            : _ptr( local._ptr )
+        {
+        }
+
+        std::shared_ptr< wait_state_t > still_alive() const { return _ptr.lock(); }
+
+        // Wake up the local function (which is using wait_until(), presumable) with a new
+        // T value
+        void signal( T const& t ) const
+        {
+            if( auto wait_state = still_alive() )
+                wait_state->signal( t );
+        }
+    };
+
+public:
+    waiting_on()
+        : _ptr( std::make_shared< wait_state_t >() )
+    {
+    }
+    waiting_on( T const & value )
+        : _ptr( std::make_shared< wait_state_t >( value ) )
+    {
+    }
+
+    // Convert to the in-thread representation
+    in_thread_ in_thread() const { return in_thread_( *this ); }
+
+    operator T const &() const { return _ptr->_value; }
+    
+    // struct value_t { double x; int k; };
+    // waiting_on< value_t > output({ 1., -1 });
+    // output->x = 2.;
+    T * operator->() { return &_ptr->_value; }
+    T const * operator->() const { return &_ptr->_value; }
+
+    // Wait until either the timeout occurs, or the predicate evaluates to true.
+    // Equivalent to:
+    //     while( ! pred() )
+    //     {
+    //         wait( timeout );
+    //         if( timed-out )
+    //             break;
+    //     }
+    template < class U, class L >
+    void wait_until( U const& timeout, L const& pred )
+    {
+        // Note that the mutex is useless and used only for the wait -- we assume here that access
+        // to the T data does not need mutual exclusion
+        std::mutex m;
+        // Following will issue (from CppCheck):
+        //     warning: The lock is ineffective because the mutex is locked at the same scope as the mutex itself. [localMutex]
+        std::unique_lock< std::mutex > locker( m );

```

Comment:
I think you need to lock the mutex on signal() too

Reply:
It's up to the user to write the lambda with concurrency in mind, if that's what's needed (for example, in dispatcher.cpp it's not needed)